content/post/2021-07-29-caching/index.Rmd

```diff
@@ -0,0 +1,223 @@
+---
+slug: cache
+title: "How to cache results in R code" 
+authors: 
+- Maëlle Salmon 
+- Christophe Dervieux
+date: "2021-07-29" 
+tags: 
+- package development 
+output: hugodown::hugo_document
+---
+
+```{r setup, include=FALSE}
+
+knitr::opts_chunk$set(fig.path = "", comment = "")
+# knitr hook to make images output use Hugo options
+knitr::knit_hooks$set(
+  plot = function(x, options) {
+    hugoopts <- options$hugoopts
+    paste0(
+      "{{<figure src=",
+      '"', x, '" ',
+      if (!is.null(hugoopts)) {
+        glue::glue_collapse(
+          glue::glue('{names(hugoopts)}="{hugoopts}"'),
+          sep = " "
+        )
+      },
+      ">}}\n"
+    )
+  }
+)
+
+```
+
+One principle of programming that's often encountered is "DRY", "Don't Repeat Yourself", that encourages e.g. the use of functions other duplicated (read: copy-pasted and slightly amended) code.
+You could also interpret it as don't let the machine repeat its calculations if useless.
+How about for a piece of code (function) with the same inputs, we only run it once per R session, and save the results for later?
+In this post, we shall go over ways to cache results in an R session, so that you don't need to burden machines.
+We will not cover [caching for R Markdown](https://bookdown.org/yihui/rmarkdown-cookbook/cache.html).
+
+## Caching: what is it and why use it?
+
+Caching means that if you call a function several times with the exact same input, the function is only actually run the first time.
+The result is stored in a cache of some sort (more practical details later!).
+Every other time the function is called with the same input, the result is retrieved from the cache.
+You will often think of caching as something valid in only one R session, but we'll see it can last longer via storage on disk.
+
+Now, *why* use caching?
+
+-   It might help save *time*.
+
+-   It might help save *other resources of users such as money*: e.g. if the function calls a web API whose pricing depends on the number of hits.
+    :sweat_smile:
+
+-   It might be *more polite*.
+    That's similar to the second item but from the perspective of e.g. a web API you keep hitting when you could have saved the result.
+    The [polite package](https://dmi3kno.github.io/polite/) for polite webscraping caches results.
+
```

Comment:
```suggestion

- Your function may asked some user inputs that won't change for the session and you don't want to ask everytime you need it. It could be per session caching but also persistent caching. As an example, **reticulate** will asked you once if you want to install miniconda by storing your answer locally if you say no and not ask again. (See internal [`miniconda_install_prompt()`](https://github.com/rstudio/reticulate/blob/edc22999925fd47e47c89e7196001446aec23806/R/miniconda.R#L284)`
```

There is also `renv::consent()` but the caching is done by checking if the action (creating renv root folder) has already be done or not. Because if you don't consent, it won't create the root env in the first place. 

Anyway, you see the idea: Caching user input is for me one important use case too. Maybe **usethis** has some examples.